lab4/lab4.rst

```diff
@@ -0,0 +1,230 @@
+There requests may be sent multiple in a single *tick* or cycle.
+
+The test that is run is initialized in ``Processor::createRecords()``.
+This function populates a queue which is the "trace" of memory accesses.
+See ``processor.cc`` for details.
+
+Feel free to add your own tests and/or modify the tests in ``createRecords``.
+When we test your code, we will use a different implementation of ``Processor::createRecords``.
+
+Cache
+~~~~~
+
+You will be inheriting from the ``Cache`` object.
+See the code in ``cache.hh`` for details of the interface you will implement.
+You will need to implement the following functions:
+- ``receiveRequest``: Called by the Processor when sending a new request.
+- ``receiveMemResponse``: Called by memory when it has finished reading the data previously requested by the cache.
+
+**DO NOT MODIFY THIS INTERFACE OR THE IMPLEMENTATION OF THE NON-VIRTUAL FUNCTIONS.**
+
+Memory
+~~~~~~
+
+Memory receives requests from the cache.
+After some delay, it will respond to the request.
+The delay is guaranteed to be more than 1 tick.
+This object also has an interface for you to determine the line size.
+The memory line size is used as the cache line size.
+
+TickedObject
+~~~~~~~~~~~~
+
+This object implements the discrete event simulation.
+Feel free to ignore it :).
+
+DirectMappedCache
+~~~~~~~~~~~~~~~~~
+
+This is an example cache implementation.
+Feel free to base you implementation of the set-associative cache on the direct mapped cache that is included.
+You can use any code from the class in your set-associative cache.
+
+TagArray and SRAMArray
+~~~~~~~~~~~~~~~~~~~~~~
+
+**DO NOT MODIFY THIS CODE IN ANY WAY**.
+
+You are required to use the ``TagArray`` and ``SRAMArray`` to hold tags/state and data, respectively.
+Constraining yourself to this interface will help ensure your cache controller design is realistic.
+
+The ``TagArray`` holds both tag and the state for each line.
+There is only room for a single tag and state for each line, so you will have to consider how to implement a set-associative cache.
+The ``TagArray`` takes the number of tag bits and data bits as parameters.
+We will check to make sure you use the correct number of tag and data bits.
+
+The ``SRAMArray`` is a simple data storage array that only takes the size of the data (cache block) and the number of elements to store as parameters.
+You will instantiate multiple tag and SRAM arrays for your set-associative cache implementation.
+
+Implementing a Set-Associative Cache
+------------------------------------
+
+You are given a file ``set_assoc.cc`` which has empty functions for each function you are required to implement.
+You will be modifying and turning in this file and ``set_assoc.hh``.
+Feel free to add new private/protected functions to the ``SetAssociativeCache``, but **DO NOT MODIFY THE PUBLIC INTERFACE**.
```

Comment:
Usually I think it's easier to enforce these things by giving them a reference header with a virtual class they have to inherit from, instead of having to check every header they submit this way.

Reply:
I'm doing that. SetAssociativeCache inherits from Cache. We won't have to check their header files, but I think it's a good idea to remind them :)